inference/server/oasst_inference_server/routes/workers.py

```diff
@@ -85,191 +72,319 @@ def get_work_request_container(work_request_map: WorkRequestContainerMap, reques
 
 
-@router.websocket("/work")
-async def handle_worker(
-    websocket: fastapi.WebSocket,
-    api_key: str = worker_utils.api_key_header,
-    protocol_version: str = worker_utils.protocol_version_header,
-):
-    await websocket.accept()
+class HandleWorkerContext(NamedTuple):
+    """
+    Whenever a worker connects to the inference server over a websocket, we create an instance of this class to hold all
+    state that is associated with that worker and connection.
+    """
 
-    try:
-        worker_utils.get_protocol_version(protocol_version)
-        api_key = worker_utils.get_api_key(api_key)
-        worker_id = await worker_utils.get_worker_id(api_key=api_key, protocol_version=protocol_version)
-    except fastapi.HTTPException as e:
-        logger.warning(f"handle_worker: {e.status_code=} {e.detail=}")
-        if e.status_code == fastapi.status.HTTP_426_UPGRADE_REQUIRED:
-            await worker_utils.send_worker_request(websocket=websocket, request=inference.UpgradeProtocolRequest())
-        elif e.status_code == fastapi.status.HTTP_401_UNAUTHORIZED:
-            await worker_utils.send_worker_request(websocket=websocket, request=inference.WrongApiKeyRequest())
-        try:
-            await websocket.close(code=e.status_code, reason=e.detail)
-        except Exception:
-            pass
-        raise fastapi.WebSocketException(e.status_code, e.detail)
-
-    logger.info(f"handle_worker: {worker_id=}")
-    worker_info = await worker_utils.receive_worker_info(websocket)
-    logger.info(f"handle_worker: {worker_info=}")
-    worker_config = worker_info.config
-    worker_compat_hash = worker_config.compat_hash
-    work_queue = queueing.work_queue(deps.redis_client, worker_compat_hash)
-    redis_client = deps.make_redis_client()
-    blocking_work_queue = queueing.work_queue(redis_client, worker_compat_hash)
-    worker_session = worker_utils.WorkerSession(
-        worker_id=worker_id,
-        worker_info=worker_info,
-    )
-    work_request_map: dict[str, WorkRequestContainer] = {}
-    pending_futures = set()
-    try:
-        async with deps.manual_create_session() as session:
-            await add_worker_connect_event(session=session, worker_id=worker_id, worker_info=worker_info)
-        await worker_utils.store_worker_session(worker_session)
+    websocket: fastapi.WebSocket
+    worker_id: str
+    worker_info: inference.WorkerInfo
 
-        async def _update_session(metrics: inference.WorkerMetricsInfo):
-            worker_session.requests_in_flight = len(work_request_map)
-            if metrics:
-                worker_session.metrics = metrics
-            await worker_utils.store_worker_session(worker_session)
+    work_queue: queueing.RedisQueue
 
-        def _add_dequeue(ftrs: set):
-            requests_in_progress = len(work_request_map)
-            if requests_in_progress < worker_config.max_parallel_requests:
-                ftrs.add(asyncio.ensure_future(blocking_work_queue.dequeue(timeout=0)))
```

Comment:
Moved to `FuturesManager.ensure_listening_for_work_requests()`. The future is now wrapped such that it returns a `tuple[FutureType, X]` where `X` is the return type of whatever the future returned before.